feat(node): added waku node builder type

LNSD opened the pull request:
This PR introduces a `WakuNode` compatible builder. The idea is to establish here the builder type and then evolve it in future iterations. This is just the first PR introducing the type and the API. It maps the `WakuNode.new` to the `WakuNodeBuilder` API.

- [x] Added the `WakuNodeBuilder` type and methods.
- [x] Added a set of `WakuNode.new` compatible `withXXX()` methods.
- [x] Moved all apps, tools, and examples to the builder pattern.

Ivansete-status commented on waku/v2/node/builder.nim:
```diff
@@ -0,0 +1,167 @@
+## Waku discv5
+
+proc withWakuDiscv5*(builder: var WakuNodeBuilder, instance: WakuDiscoveryV5) =
+  if not instance.isNil():
```
Should we do anything for the case 'isNil' ?

LNSD replied:
I added this for compatibility reasons (with the deprecated `WakuNode.new()` API). Accepting a `nil` makes 8 test cases crash 😅

alrevuelta commented on waku/v2/node/builder.nim:
```diff
@@ -0,0 +1,167 @@
+
+## Peer storage and peer manager
+
+proc withPeerStorage*(builder: var WakuNodeBuilder, peerStorage: PeerStorage, capacity = none(int)) =
```
perhaps merge with withPeerManager?

LNSD replied:
Yes! This is the idea. It will be merged after we can change the `WakuNode.new()` proc without having to change 120 test cases :)

alrevuelta reviewed (approved):
lgtm! really like the pattern

LNSD commented:
Just for the record, this PR is an intermediate step of a more extensive refactoring. 

The builder APIs here are not final. Some of the APIs will change after the `WakuNode` procedures change.

LNSD commented:
I have added some changes to the PR that will make it easier to understand than in the vacuum:

- [x] Moved all apps, tools, and examples to the builder pattern.

jm-clius reviewed (approved):
LGTM! Great improvement. Thanks.

richard-ramos commented:
Cool PR! :100:
